File: bindings/java/src/test/java/com/deftio/triepack/FixturesTest.java
/*
 * FixturesTest.java
 *
 * Cross-language fixture tests -- mirrors Python test_fixtures.py.
 *
 * Copyright (c) 2026 M. A. Chatterjee
 * SPDX-License-Identifier: BSD-2-Clause
 */

package com.deftio.triepack;

import static org.junit.jupiter.api.Assertions.*;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.LinkedHashMap;
import java.util.Map;

import org.junit.jupiter.api.Test;

class FixturesTest {

    /**
     * Locate the fixtures directory relative to this source file.
     * Works when running from bindings/java/ with gradle.
     */
    private static Path fixtureDir() {
        // Try multiple possible locations
        Path[] candidates = {
            Paths.get("../../tests/fixtures"),
            Paths.get("tests/fixtures"),
            Paths.get("../../../tests/fixtures"),
        };
        for (Path p : candidates) {
            if (Files.isDirectory(p)) {
                return p;
            }
        }
        // Fallback to absolute path
        return Paths.get(System.getProperty("user.dir"))
                     .resolve("../../tests/fixtures").normalize();
    }

    private static byte[] readFixture(String name) throws IOException {
        return Files.readAllBytes(fixtureDir().resolve(name));
    }

    // ── Decode C-generated fixtures ──────────────────────────────

    @Test
    void testDecodeEmpty() throws IOException {
        byte[] buf = readFixture("empty.trp");
        Map<String, TpValue> result = TriePack.decode(buf);
        assertTrue(result.isEmpty());
    }

    @Test
    void testDecodeSingleNull() throws IOException {
        byte[] buf = readFixture("single_null.trp");
        Map<String, TpValue> result = TriePack.decode(buf);
        assertEquals(1, result.size());
        assertEquals(TpValue.Type.NULL, result.get("hello").getType());
    }

    @Test
    void testDecodeSingleInt() throws IOException {
        byte[] buf = readFixture("single_int.trp");
        Map<String, TpValue> result = TriePack.decode(buf);
        assertEquals(1, result.size());
        assertEquals(42, result.get("key").uintValue());
    }

    @Test
    void testDecodeMultiMixed() throws IOException {
        byte[] buf = readFixture("multi_mixed.trp");
        Map<String, TpValue> result = TriePack.decode(buf);
        assertTrue(result.get("bool").boolValue());
        assertEquals(3.14159, result.get("f64").float64Value(), 1e-10);
        assertEquals(-100, result.get("int").intValue());
        assertEquals("hello", result.get("str").stringValue());
        assertEquals(200, result.get("uint").uintValue());
    }

    @Test
    void testDecodeSharedPrefix() throws IOException {
        byte[] buf = readFixture("shared_prefix.trp");
        Map<String, TpValue> result = TriePack.decode(buf);
        assertEquals(10, result.get("abc").uintValue());
        assertEquals(20, result.get("abd").uintValue());
        assertEquals(30, result.get("xyz").uintValue());
    }

    @Test
    void testDecodeLarge() throws IOException {
        byte[] buf = readFixture("large.trp");
        Map<String, TpValue> result = TriePack.decode(buf);
        assertEquals(100, result.size());
        for (int i = 0; i < 100; i++) {
            String key = String.format("key_%04d", i);
            assertEquals(i, result.get(key).uintValue());
        }
    }

    @Test
    void testDecodeKeysOnly() throws IOException {
        byte[] buf = readFixture("keys_only.trp");
        Map<String, TpValue> result = TriePack.decode(buf);
        assertEquals(3, result.size());
        assertEquals(TpValue.Type.NULL, result.get("apple").getType());
        assertEquals(TpValue.Type.NULL, result.get("banana").getType());
        assertEquals(TpValue.Type.NULL, result.get("cherry").getType());
    }

    // ── Byte-for-byte binary reproducibility ─────────────────────

    @Test
    void testEncodeEmptyMatchesFixture() throws IOException {
        Map<String, TpValue> data = new LinkedHashMap<>();
        byte[] javaBuf = TriePack.encode(data);
        byte[] cBuf = readFixture("empty.trp");
        assertArrayEquals(cBuf, javaBuf);
    }

    @Test
    void testEncodeSingleNullMatchesFixture() throws IOException {
        Map<String, TpValue> data = new LinkedHashMap<>();
        data.put("hello", TpValue.ofNull());
        byte[] javaBuf = TriePack.encode(data);
        byte[] cBuf = readFixture("single_null.trp");
        assertArrayEquals(cBuf, javaBuf);
    }

    @Test
    void testEncodeSingleIntMatchesFixture() throws IOException {
        Map<String, TpValue> data = new LinkedHashMap<>();
        data.put("key", TpValue.ofUInt(42));
        byte[] javaBuf = TriePack.encode(data);
        byte[] cBuf = readFixture("single_int.trp");
        assertArrayEquals(cBuf, javaBuf);
    }

    @Test
    void testEncodeMultiMixedMatchesFixture() throws IOException {
        Map<String, TpValue> data = new LinkedHashMap<>();
        data.put("bool", TpValue.ofBool(true));
        data.put("f64", TpValue.ofFloat64(3.14159));
        data.put("int", TpValue.ofInt(-100));
        data.put("str", TpValue.ofString("hello"));
        data.put("uint", TpValue.ofUInt(200));
        byte[] javaBuf = TriePack.encode(data);
        byte[] cBuf = readFixture("multi_mixed.trp");
        assertArrayEquals(cBuf, javaBuf);
    }

    @Test
    void testEncodeSharedPrefixMatchesFixture() throws IOException {
        Map<String, TpValue> data = new LinkedHashMap<>();
        data.put("abc", TpValue.ofUInt(10));
        data.put("abd", TpValue.ofUInt(20));
        data.put("xyz", TpValue.ofUInt(30));
        byte[] javaBuf = TriePack.encode(data);
        byte[] cBuf = readFixture("shared_prefix.trp");
        assertArrayEquals(cBuf, javaBuf);
    }

    @Test
    void testEncodeLargeMatchesFixture() throws IOException {
        Map<String, TpValue> data = new LinkedHashMap<>();
        for (int i = 0; i < 100; i++) {
            data.put(String.format("key_%04d", i), TpValue.ofUInt(i));
        }
        byte[] javaBuf = TriePack.encode(data);
        byte[] cBuf = readFixture("large.trp");
        assertArrayEquals(cBuf, javaBuf);
    }

    @Test
    void testEncodeKeysOnlyMatchesFixture() throws IOException {
        Map<String, TpValue> data = new LinkedHashMap<>();
        data.put("apple", TpValue.ofNull());
        data.put("banana", TpValue.ofNull());
        data.put("cherry", TpValue.ofNull());
        byte[] javaBuf = TriePack.encode(data);
        byte[] cBuf = readFixture("keys_only.trp");
        assertArrayEquals(cBuf, javaBuf);
    }
}

File: bindings/kotlin/src/test/kotlin/com/deftio/triepack/FixturesTest.kt
/*
 * FixturesTest.kt
 *
 * Cross-language fixture tests -- mirrors Python test_fixtures.py.
 *
 * Copyright (c) 2026 M. A. Chatterjee
 * SPDX-License-Identifier: BSD-2-Clause
 */

package com.deftio.triepack

import java.io.File
import kotlin.math.abs
import kotlin.test.Test
import kotlin.test.assertContentEquals
import kotlin.test.assertEquals
import kotlin.test.assertNotNull
import kotlin.test.assertNull
import kotlin.test.assertTrue

class FixturesTest {

    private fun fixtureDir(): String {
        // Navigate from bindings/kotlin/ up to tests/fixtures/
        val candidates = listOf(
            // Running from project root
            "tests/fixtures",
            // Running from bindings/kotlin
            "../../tests/fixtures",
            // Absolute path fallback
            System.getProperty("user.dir") + "/../../tests/fixtures"
        )
        for (c in candidates) {
            val f = File(c)
            if (f.exists() && f.isDirectory) return f.absolutePath
        }
        // Last resort: walk up from CWD
        var dir = File(System.getProperty("user.dir"))
        while (dir.parentFile != null) {
            val fixture = File(dir, "tests/fixtures")
            if (fixture.exists() && fixture.isDirectory) return fixture.absolutePath
            dir = dir.parentFile
        }
        throw RuntimeException("Cannot find tests/fixtures directory")
    }

    private fun readFixture(name: String): ByteArray {
        return File(fixtureDir(), name).readBytes()
    }

    // -- Decode C-generated fixtures --

    @Test
    fun testDecodeEmpty() {
        val buf = readFixture("empty.trp")
        val result = decode(buf)
        assertEquals(emptyMap(), result)
    }

    @Test
    fun testDecodeSingleNull() {
        val buf = readFixture("single_null.trp")
        val result = decode(buf)
        assertEquals(1, result.size)
        assertNull(result["hello"])
    }

    @Test
    fun testDecodeSingleInt() {
        val buf = readFixture("single_int.trp")
        val result = decode(buf)
        assertEquals(TpValue.UInt(42), result["key"])
    }

    @Test
    fun testDecodeMultiMixed() {
        val buf = readFixture("multi_mixed.trp")
        val result = decode(buf)
        assertEquals(TpValue.Bool(true), result["bool"])
        val f64 = result["f64"]
        assertNotNull(f64)
        assertTrue(f64 is TpValue.Float64)
        assertTrue(abs(f64.value - 3.14159) < 1e-10)
        assertEquals(TpValue.Int(-100), result["int"])
        assertEquals(TpValue.Str("hello"), result["str"])
        assertEquals(TpValue.UInt(200), result["uint"])
    }

    @Test
    fun testDecodeSharedPrefix() {
        val buf = readFixture("shared_prefix.trp")
        val result = decode(buf)
        assertEquals(TpValue.UInt(10), result["abc"])
        assertEquals(TpValue.UInt(20), result["abd"])
        assertEquals(TpValue.UInt(30), result["xyz"])
    }

    @Test
    fun testDecodeLarge() {
        val buf = readFixture("large.trp")
        val result = decode(buf)
        assertEquals(100, result.size)
        for (i in 0 until 100) {
            val key = "key_%04d".format(i)
            assertEquals(TpValue.UInt(i.toLong()), result[key])
        }
    }

    @Test
    fun testDecodeKeysOnly() {
        val buf = readFixture("keys_only.trp")
        val result = decode(buf)
        assertEquals(3, result.size)
        assertNull(result["apple"])
        assertNull(result["banana"])
        assertNull(result["cherry"])
    }

    // -- Byte-for-byte binary reproducibility --

    @Test
    fun testEncodeEmptyMatchesFixture() {
        val ktBuf = encode(emptyMap())
        val cBuf = readFixture("empty.trp")
        assertContentEquals(cBuf, ktBuf)
    }

    @Test
    fun testEncodeSingleNullMatchesFixture() {
        val ktBuf = encode(mapOf("hello" to null))
        val cBuf = readFixture("single_null.trp")
        assertContentEquals(cBuf, ktBuf)
    }

    @Test
    fun testEncodeSingleIntMatchesFixture() {
        val ktBuf = encode(mapOf("key" to TpValue.UInt(42)))
        val cBuf = readFixture("single_int.trp")
        assertContentEquals(cBuf, ktBuf)
    }

    @Test
    fun testEncodeMultiMixedMatchesFixture() {
        val ktBuf = encode(mapOf(
            "bool" to TpValue.Bool(true),
            "f64" to TpValue.Float64(3.14159),
            "int" to TpValue.Int(-100),
            "str" to TpValue.Str("hello"),
            "uint" to TpValue.UInt(200)
        ))
        val cBuf = readFixture("multi_mixed.trp")
        assertContentEquals(cBuf, ktBuf)
    }

    @Test
    fun testEncodeSharedPrefixMatchesFixture() {
        val ktBuf = encode(mapOf(
            "abc" to TpValue.UInt(10),
            "abd" to TpValue.UInt(20),
            "xyz" to TpValue.UInt(30)
        ))
        val cBuf = readFixture("shared_prefix.trp")
        assertContentEquals(cBuf, ktBuf)
    }

    @Test
    fun testEncodeLargeMatchesFixture() {
        val data = mutableMapOf<String, TpValue?>()
        for (i in 0 until 100) {
            data["key_%04d".format(i)] = TpValue.UInt(i.toLong())
        }
        val ktBuf = encode(data)
        val cBuf = readFixture("large.trp")
        assertContentEquals(cBuf, ktBuf)
    }

    @Test
    fun testEncodeKeysOnlyMatchesFixture() {
        val ktBuf = encode(mapOf(
            "apple" to null,
            "banana" to null,
            "cherry" to null
        ))
        val cBuf = readFixture("keys_only.trp")
        assertContentEquals(cBuf, ktBuf)
    }
}
